pandas/indexes/interval.py

```diff
@@ -0,0 +1,983 @@
+            # slice
+            start, stop = key.start, key.stop
+            if (key.step or 1) != 1:
+                raise NotImplementedError("cannot slice with a slice step")
+            if start is None:
+                start = 0
+            else:
+                start = self._searchsorted_monotonic(start, 'left')
+            if stop is None:
+                stop = len(self)
+            else:
+                stop = self._searchsorted_monotonic(stop, 'right')
+        else:
+            # scalar or index-like
+
+            start = self._searchsorted_monotonic(key, 'left')
+            stop = self._searchsorted_monotonic(key, 'right')
+        return start, stop
+
+    def get_loc(self, key, method=None):
+        self._check_method(method)
+
+        original_key = key
+        key = self._maybe_cast_indexed(key)
+
+        if self.is_non_overlapping_monotonic:
+            if isinstance(key, Interval):
+                left = self._maybe_cast_slice_bound(key.left, 'left', None)
+                right = self._maybe_cast_slice_bound(key.right, 'right', None)
+                key = Interval(left, right, key.closed)
+            else:
+                key = self._maybe_cast_slice_bound(key, 'left', None)
+
+            start, stop = self._find_non_overlapping_monotonic_bounds(key)
+
+            if start is None or stop is None:
+                return slice(start, stop)
+            elif start + 1 == stop:
+                return start
+            elif start < stop:
+                return slice(start, stop)
+            else:
+                raise KeyError(original_key)
+
+        else:
+            # use the interval tree
+            if isinstance(key, Interval):
+                left, right = _get_interval_closed_bounds(key)
+                return self._engine.get_loc_interval(left, right)
+            else:
+                return self._engine.get_loc(key)
+
+    def get_value(self, series, key):
+        if com.is_bool_indexer(key):
+            loc = key
+        elif is_list_like(key):
+            loc = self.get_indexer(key)
+        else:
+            loc = self.get_loc(key)
+        return series.iloc[loc]
+
+    @Appender(_index_shared_docs['get_indexer'] % _index_doc_kwargs)
+    def get_indexer(self, target, method=None, limit=None, tolerance=None):
+        self._check_method(method)
+        target = _ensure_index(target)
+        target = self._maybe_cast_indexed(target)
+
+        if self.equals(target):
+            return np.arange(len(self), dtype='intp')
+
+        if self.is_non_overlapping_monotonic:
+            start, stop = self._find_non_overlapping_monotonic_bounds(target)
+
+            start_plus_one = start + 1
+            if not ((start_plus_one < stop).any()):
+                return np.where(start_plus_one == stop, start, -1)
+
+        if not self.is_unique:
+            raise ValueError("get_indexer cannot handle non-unique indices")
+
+        # find the left and right indexers
```

Comment:
This is the crux of the: 'how do we handle overlaps' with multiple intervals. Maybe we shouldn't even allow this (ATM). And just not support a *list-of-Intervals* at all as an indexer. Is there a use-case for this?